core/lib/multivm/src/versions/era_vm/tracers/pubdata_tracer.rs

```diff
@@ -0,0 +1,172 @@
+use itertools::Itertools;
+use zksync_state::ReadStorage;
+use zksync_types::{
+    event::{
+        extract_l2tol1logs_from_l1_messenger, extract_long_l2_to_l1_messages,
+        L1_MESSENGER_BYTECODE_PUBLICATION_EVENT_SIGNATURE,
+    },
+    writes::StateDiffRecord,
+    AccountTreeId, StorageKey, L1_MESSENGER_ADDRESS, U256,
+};
+use zksync_utils::u256_to_h256;
+
+use super::traits::{Tracer, Vm, VmTracer};
+use crate::{
+    era_vm::{
+        bootloader_state::utils::{apply_pubdata_to_memory, PubdataInput},
+        event::merge_events,
+        hook::Hook,
+    },
+    interface::tracer::{TracerExecutionStatus, TracerExecutionStopReason},
+    vm_1_4_1::VmExecutionMode,
+};
+
+pub struct PubdataTracer {
+    execution_mode: VmExecutionMode,
+    pubdata_before_run: i32,
+    should_stop: bool,
+    pub pubdata_published: u32,
+    // this field is to enforce a custom storage diff when setting the pubdata to the bootloader
+    // this is meant to be used for testing purposes only.
+    enforced_storage_diff: Option<Vec<StateDiffRecord>>,
+}
+
+impl PubdataTracer {
+    pub fn new(execution_mode: VmExecutionMode) -> Self {
+        Self {
+            execution_mode,
+            pubdata_before_run: 0,
+            pubdata_published: 0,
+            enforced_storage_diff: None,
+            should_stop: false,
+        }
+    }
+
+    pub fn new_with_forced_state_diffs(
+        execution_mode: VmExecutionMode,
+        diff: Vec<StateDiffRecord>,
+    ) -> Self {
+        Self {
+            enforced_storage_diff: Some(diff),
+            ..Self::new(execution_mode)
+        }
+    }
+
+    fn get_storage_diff<S: ReadStorage>(&mut self, vm: &mut Vm<S>) -> Vec<StateDiffRecord> {
+        vm.inner
+            .state
+            .get_storage_changes(&mut vm.world)
+            .iter()
+            .filter_map(|(storage_key, initial_value, value)| {
+                let address = storage_key.address;
+
+                if address == L1_MESSENGER_ADDRESS {
+                    return None;
+                }
+
+                let key = storage_key.key;
+
+                let diff = StateDiffRecord {
+                    key,
+                    address,
+                    derived_key:
+                        zk_evm_1_5_0::aux_structures::LogQuery::derive_final_address_for_params(
+                            &address, &key,
+                        ),
+                    enumeration_index: vm
+                        .storage
+                        .borrow_mut()
```

Comment:
I don't like borrowing refcells outside of the implant block for this struct:
The reason is that this will panic in runtime if there is another living reference to it, which it could possibly happen given that this field is public, let's change this, do we need to change something else in our vm?